plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -494,67 +519,94 @@ applyBuiltin ctx bn args = do
+-- | Evaluate a term using the CEK machine and keep track of costing, logging is optional.
 runCek
     :: ( GShow uni, GEq uni, Closed uni, uni `Everywhere` ExMemoryUsage
        , Hashable fun, Ix fun, ExMemoryUsage fun
        )
     => BuiltinsRuntime fun (CekValue uni fun)
     -> ExBudgetMode
+    -> Bool
     -> Term Name uni fun ()
-    -> (Either (CekEvaluationException uni fun) (Term Name uni fun ()), CekExBudgetState fun)
-runCek runtime mode term =
-    runCekM (CekEnv runtime mode)
-            (ExBudgetState mempty mempty)
-        $ do
-            spendBudget BAST (ExBudget 0 (termAnn memTerm))
-            computeCek [] mempty memTerm
-    where
-        memTerm = withMemory term
-
--- | Evaluate a term using the CEK machine in the 'Counting' mode.
-runCekCounting
+    -> (Either (CekEvaluationException uni fun) (Term Name uni fun ()), CekExBudgetState fun, [String])
+runCek runtime mode emitting term =
+    runCekM runtime mode emitting (ExBudgetState mempty mempty) $ do
+        spendBudget BAST (ExBudget 0 (termAnn memTerm))
+        computeCek [] mempty memTerm
+  where
+    memTerm = withMemory term
+
+-- | Evaluate a term using the CEK machine with logging disabled and keep track of costing.
+runCekNoEmit
     :: ( GShow uni, GEq uni, Closed uni, uni `Everywhere` ExMemoryUsage
        , Hashable fun, Ix fun, ExMemoryUsage fun
        )
     => BuiltinsRuntime fun (CekValue uni fun)
+    -> ExBudgetMode
     -> Term Name uni fun ()
     -> (Either (CekEvaluationException uni fun) (Term Name uni fun ()), CekExBudgetState fun)
-runCekCounting runtime = runCek runtime Counting
+runCekNoEmit runtime mode term =
+    case runCek runtime mode False term of
+        (errOrRes, s', _) -> (errOrRes, s')
 
--- | Evaluate a term using the CEK machine.
+-- | Unsafely evaluate a term using the CEK machine with logging disabled and keep track of costing.
+-- May throw a 'CekMachineException'.
+unsafeRunCekNoEmit
+    :: ( GShow uni, GEq uni, Typeable uni
+       , Closed uni, uni `EverywhereAll` '[ExMemoryUsage, PrettyConst]
+       , Hashable fun, Ix fun, Pretty fun, Typeable fun, ExMemoryUsage fun
+       )
+    => BuiltinsRuntime fun (CekValue uni fun)
+    -> ExBudgetMode
+    -> Term Name uni fun ()
+    -> (EvaluationResult (Term Name uni fun ()), CekExBudgetState fun)
+unsafeRunCekNoEmit runtime mode = first unsafeExtractEvaluationResult . runCekNoEmit runtime mode
+
+-- | Evaluate a term using the CEK machine with logging enabled.
 evaluateCek
+    :: ( GShow uni, GEq uni, Closed uni, uni `Everywhere` ExMemoryUsage
+       , Hashable fun, Ix fun, ExMemoryUsage fun
+       )
+    => BuiltinsRuntime fun (CekValue uni fun)
+    -> Term Name uni fun ()
+    -> (Either (CekEvaluationException uni fun) (Term Name uni fun ()), [String])
+evaluateCek runtime term =
+    -- TODO: Oftentimes we want neither 'Restricting' nor 'Counting'. Should we have a third mode
+    -- 'JustEvaluateTheProgram'?
```

Comment:
>  I'd imagined that in cases like that we'd just run with a large initial budget

That actually makes a lot of sense. I think I like this option more than others (even with reworked budgeting), 'cause benchmarking results automatically translate to on-chain execution times (as you say) and so it's exactly what we want. With smarter versions there's still room for a screw-up.

Reply:
I was a little worried that if we started with a _really_ big budget then maybe the time taken to do arithmetic with it would itself slow the program down more than required.  I suppose that we'd mostly be doing comparisons though, so maybe that doesn't matter.  

There are (at leas) two ways we could do the on-chain cost counting: (a) add up costs as we go along and compare them with the initial budget limit, failing when we exceed it, and (b) start with an initial budget and decrement it as we go along, failing when it drops below zero.  I had a vague idea that (a) would add a bit more overhead because it'd be more expensive to compare two numbers than to check whether a number is negative, but that's probably nonsense because in realistic cases I'm sure we'll stay inside a single machine word and both of those things will take a single instruction each.

Reply:
Yeah, a do-nothing mode sounds useful.